gems/kaminari-activerecord/1.2/kaminari-activerecord.rbs

```diff
@@ -0,0 +1,29 @@
+module Kaminari
+  module ActiveRecordModelExtension
+    module ClassMethods
```

Comment:
# facts

Well, let's get the facts in order first.

In Ruby, `Kaminari::ActiveRecordModelExtension::ClassMethods` and the related classes are the following.

* `Kaminari::ActiveRecordModelExtension::ClassMethods` module does not exist
* `Kaminari::ActiveRecordModelExtension` is a module extended by `ActiveSupport::Concern` [ref](https://github.com/kaminari/kaminari/blob/bf20fccfdbf7d10c337aeead7ecf4f54a230a5fb/kaminari-activerecord/lib/kaminari/activerecord/active_record_model_extension.rb#L6-L7)
* On `included` hook of `ActiveRecordModelExtension`
  * `Kaminari::ConfigurationMethods` is included [ref](https://github.com/kaminari/kaminari/blob/bf20fccfdbf7d10c337aeead7ecf4f54a230a5fb/kaminari-activerecord/lib/kaminari/activerecord/active_record_model_extension.rb#L10)
    * `ClassMethods` of this module defines `paginates_per` and so on.
  * `page` class method is defined as the included class. (The name depends on the configuration) [ref](https://github.com/kaminari/kaminari/blob/bf20fccfdbf7d10c337aeead7ecf4f54a230a5fb/kaminari-activerecord/lib/kaminari/activerecord/active_record_model_extension.rb#L14-L22)
* `Kaminari::ActiveRecordModelExtension` is included to subclasses of `::ActiveRecord::Base` [ref](https://github.com/kaminari/kaminari/blob/bf20fccfdbf7d10c337aeead7ecf4f54a230a5fb/kaminari-activerecord/lib/kaminari/activerecord/active_record_extension.rb#L13-L20)


So, the inheritance is the following in Ruby (When a `User` class is defined).


```ruby
module Kaminari
  module ActiveRecordModelExtension
    extend ActiveSupport::Concern
  end
end

class ApplicationRecord < ActiveRecord::Base
  include Kaminari::ActiveRecordModelExtension # by inherited/included hook of ActiveRecordExtension
  include Kaminari::ConfigurationMethods # by included hook of Kaminari::ActiveRecordModelExtension
  include Kaminari::ConfigurationMethods::ClassMethods # by ActiveSupport::Concern

  def self.page(...) # by included hook of Kaminari::ActiveRecordModelExtension
    # ...
  end
end

class User < ApplicationRecord
end
```

# In RBS

Basically, I think the type definitions should follow Ruby's structure.
But there are problems if we define kaminari-acitverecord's RBS as the same as Ruby.


* Users need to write all `include` and `extend` to a class(es) that inherits `AcitveRecord::Base`.
* Users need to define the `page` class method by themselves.

Your first approach solves these problems by introducing pseudo `Kaminari::ActiveRecordModelExtension::ClassMethods` module. It reduces the first problem and solves the second problem. But the inheritance structure is different from Ruby.
I'm worried that the difference confuses the user.


I'm thinking of just adding an interface to define `page` method. For example:

```ruby
# kaminari-activerecord.rbs

module Kaminari
  module ActiveRecordModelExtension
    interface _ClassMethods
      def page: (blah) -> blah
    end

    # ...
  end
end

# RBS file in user's code base

class ApplicationRecord < ActiveRecord::Base
  include Kaminari::ActiveRecordModelExtension
  include Kaminari::ConfigurationMethods
  include Kaminari::ConfigurationMethods::ClassMethods

  extend Kaminari::ActiveRecordModelExtension::_ClassMethods
end

class User < ApplicationRecord
end
```

It solves the second problem. Users just need to write `extend Kaminari::ActiveRecordModelExtension::_ClassMethods` to define `page` method.
It doesn't solve the first problem. Personally I think it is acceptable because it's hard to keep the convenience and right inheritance. I hope this problem will be solved by another layer, such as the RBS runtime or a generator. So I think we can leave this problem for now.